### DISCUSSION TITLE
When to use these implementations of GraphQL and Next.js

### DISCUSSION THREAD

**Discussion by vitro-dynamics** [vercel/next.js#16201] [Help]
Howdy,

I was checking out some GraphQL implementations with Next.js. The implementations in question are as follows: 

1.) If you go to the examples folder on Github you can find this [`with-apollo`](https://github.com/vercel/next.js/tree/canary/examples/with-apollo) example. It looks like it is SSR with an implementation with Apollo Client.  

2.) My guy, Ben Grunfeld, implements GraphQL on the the Next.js API routes using Apollo Server. [Link here](https://medium.com/swlh/server-side-graphql-with-apollo-nextjs-part-1-setup-2615410c4966). 

I guess my question is when would I go with one implementation over the other? I can only surmise that it depends on what I'm using as a data source. Like for example, If I used Hasura, I'd use the first implementation, as compared to using Firebase, which at which point I'd use the second implementation (cause I'd need to code the GraphQL interaction with Firebase). 

does anyone have further insight on this? I'd love to hear it.

**Comment by leerob** [comment]
Great question. In general, I'd say it boils down to these two use cases:

1. Your GraphQL API already exists (e.g., [Hasura](https://github.com/leerob/hasura-nextjs-ssr)) and you want to fetch some data from it. You could use Apollo, graphql-request or vanilla fetch to retrieve this data.
2. You want to _build_ a GraphQL API with Next.js using serverless functions to read from some database. In this case, you would use API routes.

If you're using Firebase, you likely wouldn't use GraphQL at all and instead [use their SDKs](https://github.com/leerob/nextjs-vercel-firebase).

**Reply by vitro-dynamics** [reply]
Lee, whats up my guy! Thank you for phrasing that way better than me lol. This has been noted!
